rye/src/cli/publish.rs

```diff
@@ -181,30 +194,161 @@ pub fn execute(cmd: Args) -> Result<(), Error> {
+        credentials.username = Some(DEFAULT_USERNAME.to_string())
+    }
+
+    credentials
+}
+
+fn resolve_repository(
+    credentials_table: Option<&Item>,
+    name: Option<String>,
+    url: Option<Url>,
+) -> Result<Repository, Error> {
+    let mut repository = Repository { name, url };
+
+    if repository.url.is_some() {
+        return Ok(repository);
+    }
+
+    if let Some(cred_url) = credentials_table.and_then(|it| {
+        it.get("repository-url")
+            .map(Item::to_string)
+            .map(escape_string)
+    }) {
+        repository.url = Some(Url::parse(&cred_url)?);
+    }
+
+    if repository.url.is_none()
+        && repository
+            .name
+            .as_ref()
+            .map_or(false, |it| it == DEFAULT_REPOSITORY)
+    {
+        repository.url = Some(Url::parse(DEFAULT_REPOSITORY_URL)?);
+    }
+
+    Ok(repository)
+}
+
+fn prompt_repository_url() -> Result<Option<Url>, Error> {
+    eprint!("Repository URL: ");
+    let url = get_trimmed_user_input().context("failed to read provided url")?;
+
+    if url.is_empty() {
+        Ok(None)
+    } else {
+        Ok(Some(Url::parse(&url)?))
+    }
+}
+
+fn prompt_token() -> Result<Option<Secret<String>>, Error> {
     eprint!("Access token: ");
+    // TODO(cnpryer): Use dialoguer?
     let token = get_trimmed_user_input().context("failed to read provided token")?;
 
-    Ok(token)
+    if token.is_empty() {
+        Ok(None)
+    } else {
+        Ok(Some(Secret::new(token)))
+    }
 }
 
-fn maybe_encrypt(secret: &Secret<String>, yes: bool) -> Result<Secret<Vec<u8>>, Error> {
-    let phrase = if !yes {
-        dialoguer::Password::with_theme(tui_theme())
-            .with_prompt("Encrypt with passphrase (optional)")
-            .allow_empty_password(true)
-            .report(false)
-            .interact()
-            .map(Secret::new)?
+fn prompt_username() -> Result<Option<String>, Error> {
+    eprint!("Username: ");
+    let username = get_trimmed_user_input().context("failed to read provided username")?;
+
+    if username.is_empty() {
+        Ok(None)
     } else {
-        Secret::new("".to_string())
-    };
+        Ok(Some(username))
+    }
+}
 
+fn prompt_password() -> Result<Option<Secret<String>>, Error> {
+    let password = dialoguer::Password::with_theme(tui_theme())
+        .with_prompt("Password")
+        .allow_empty_password(true)
+        .report(false)
+        .interact()?;
+
+    if password.is_empty() {
+        Ok(None)
+    } else {
+        Ok(Some(Secret::new(password)))
+    }
+}
+
+fn encrypt(secret: &Secret<String>, phrase: &Secret<String>) -> Result<Secret<Vec<u8>>, Error> {
     let token = if phrase.expose_secret().is_empty() {
         secret.expose_secret().as_bytes().to_vec()
     } else {
         // Do the encryption
-        let encryptor = Encryptor::with_user_passphrase(phrase);
+        let encryptor = Encryptor::with_user_passphrase(phrase.clone());
```

Comment:
Owning `phrase` is important for resolving decrypt vs encrypt options. I want to hang onto that beyond calling this function.